Build script modifications

mscappini opened the pull request.

Commit:
Created a Grunt script to run.

aisthesis commented on package.json:
```diff
@@ -1,15 +1,19 @@
-  "dependencies": {
-    "express": "4.x.x",
-    "express-handlebars": "2.x.x"
+    "name": "investorbot",
```
Please edit appropriately. Name should be "ui". Version should be "0.0.0" or "0.0.1". It should have a description, etc. Cf. https://github.com/optbot/template/blob/master/package.json

aisthesis commented:
A few additional issues:

The `README.md` should reflect your changes as well as any additional requirements (presumably at least a populated mongoDB with an 'options' db and a 'quotes' collection?).

Will it accept connections only from localhost? If so, I won't be able to tell whether it's working or not.

Commit:
Updated package version, description, and authors.

Commit:
Now using 'npm start' to start the app.
Updated the README to reflect the new changes.

aisthesis commented on README.md:
```diff
@@ -17,9 +17,10 @@ Build
 
-        $ cd <project root>
-        $ node app.js
+        $ npm start [-- [--port=your_port] [--host=your_host]]
```
The syntax is unclear. I'm assuming the outer brackets and '--' are not supposed to be there, i.e., it wouldn't be correct to type `npm start -- --port=8080 --host=10.10.10.10`. So, shouldn't it say:

```
npm start [--port=8080] [--host=127.0.0.1]
```

If the options aren't required, the defaults should also be documented.

mscappini replied:
So, this syntax is actually correct. The FIRST double dash (as noted in line 20), is actually for telling `npm` that the following things are args passed to your script. Whatever you type after "--" are args that get passed directly, as-is to your script.

So, then, by running:

npm start -- --port=18080 --host=10.10.10.10

The script that actually runs by `npm` is:

 node app.js --port=18080 --host=10.10.10.10

It's ALL in brackets because the entire statement as args is optional. But so are each of the args that you pass individually.

aisthesis replied:
You should give a few examples in that case.

Also, is there any reasonable use case where we would want it on a host other than 127.0.0.1? This is referring to the host where the service is being started and not the host from which connections are going to be accepted, right?

mscappini replied:
It sort of depends on the situation. If we want like a load balancer or something between the internet and this host, we would just make the site bind to an intranet IP. If not, we'd just allow it to bind to its default, public address. I don't think there's any production scenario where this UI app binds to 127.0.0.1.

aisthesis commented on README.md:
```diff
@@ -17,9 +17,10 @@ Build
+2.  Navigate to the locally served page (unless overridden) at:
 
-2.  Navigate to the locally served page at: http://127.0.0.1:8080/
+        `http://127.0.0.1:8080/`
```
Isn't the intention to run this on a remote server which has access to a mongoDB instance containing the data source?

mscappini replied:
These values were actually supposed to be implied defaults. But I can change that around. What would be the preferred approach for saying something like: `http://whatever_host_you_choose_or_127.0.0.1:same_with_port_or_8080`?

aisthesis replied:
Let's consider the production use case: Installing the service on a cloud server and then using it (from wherever). What parameters do I use for that, both on `npm start` and then if, for example, I'm trying to use the service from my local machine?

mscappini replied:
Actually, not providing any parameters will, by default, cause it to [listen on all IP addresses](https://nodejs.org/api/http.html#http_server_listen_port_hostname_backlog_callback). I would think this is the preferred behavior, where we would instead override this to be 127.0.0.1 for local development? Even in that case, it's probably fine to bind to all IP addresses even for local development.

aisthesis commented on package.json:
```diff
@@ -1,15 +1,23 @@
-  "name": "investorbot",
-  "version": "1.0.0",
-  "description": "",
+  "name": "optbotui",
```
The name should just be "ui".

mscappini replied:
Should it really be? I thought the package.json was what this package is in relation to every package on the internet. If it's just called 'ui', wouldn't that be insufficient?

https://docs.npmjs.com/files/package.json#name

aisthesis replied:
Let's use scopes: https://docs.npmjs.com/misc/scope

If the identifiers have to be globally unique, then we're going to prefix them with optbot one way or another anyway. So: `@optbot/ui`.

mscappini replied:
Nice! I like it.

Commit:
Scoped the package name.

Commit:
Removed default IP for local binding. Default listening address will be any.
Removed unnecessary semicolons.

Commit:
Updated the readme with examples.

aisthesis commented on README.md:
```diff
@@ -17,9 +18,23 @@ Build
+
+**Start on a different port**
 
-2.  Navigate to the locally served page at: http://127.0.0.1:8080/
```
Don't delete the example for using the service. It would be best to supply an exact instructions given default port and some given external IP address.

mscappini replied:
I think that if you're this far in the process, you have a general understanding of how IP/port webserver works. If not, why are you here? I found that having the local address just made it more confusing. There's enough information from step 2 to know how to reach the served page.

aisthesis replied:
What is easier to understand:
1. "If the service is running on port 8080 on a host with IP address 91.198.174.192 paste `http://  91.198.174.192:8080/" into your browser" or
2. "The site can be navigated to by the IP and port that was bound. If not changed, the defaults are any IP address and port 8080."

mscappini replied:
I am fine with the instructions how they are. However, if you would like to see something in particular, instead of asking me to continue to iterate on it while you have a particular change in mind, for future reference, please provide some examples so that I can better understand what you are looking for. I can use this example that you have provided if you like.

aisthesis commented:
Could you add a 'Dependencies' section to the README.md to enumerate the dependencies? I have a comment above, but that seems to have gotten lost.

mscappini commented:
I don't understand. The dependencies come with `npm install`.

aisthesis commented:
But it won't work if there isn't at least a mongoDB (populated with quote data) to connect to, right? Does it currently also depend on restapi?

mscappini commented:
Oh, I see what you mean. Well, it was built so that those things aren't hard dependencies. This project would run without the REST API and MongoDB running, but in its current form, it wouldn't really do anything else. But it doesn't mean it would always be that way. It's up to you if you want to list those other pieces as strict dependencies right now.

aisthesis commented:
Dependencies was the first term that came to mind, but I'm happy if you can think of a better one. The README should explain what it actually is for the service to be "working" and what is needed for that to happen.

Irialad commented:
Yeah, anything that can exist on a different machine shouldn't be listed as
a dependency. At that point your talking about the overall system, so a
wiki, readme, or something of the like would be where I'd expect to find
the information.
On Apr 12, 2015 10:51, "Marshall Farrier" notifications@github.com wrote:

> Dependencies was the first term that came to mind, but I'm happy if you
> can think of a better one. The README should explain what it actually is
> for the service to be "working" and what is needed for that to happen.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/optbot/ui/pull/1#issuecomment-92095651.

aisthesis commented:
What word should we use to describe this relationship? I'm about to revise the `template` for new services, and there needs to be a placeholder for that in the `README.md`.

We also need a placeholder for a brief description of what each service does, and that description should clearly state what exactly has to happen for the service to be installed correctly. Filling in the `test` part in `package.json` would be ideal, but that's unrealistic at this point. But even if we're often doing manual testing, anyone who installs should at least be able to run through some number of copy-paste actions to determine whether or not the installation was successful.

So, maybe the `README.md` should always have a `Testing` section, which in the future might consist of the single line `npm test` but for now would just describe what you can do to see if it's working.

Maybe we could call the additional section 'Connects to' or 'Connections' to indicate that meaningful results depend on the service being able to connect to services a, b, and c? So, `template` and all other services would have sections:
- Testing
- Connects to

Thoughts?

aisthesis commented:
I'm getting errors when I try to install on my Mac:

```
optbot/ui$ npm install
npm ERR! install Couldn't read dependencies
npm ERR! Error: Invalid name: "@optbot/ui"
npm ERR!     at validName (/usr/local/lib/node_modules/npm/node_modules/read-package-json/read-json.js:550:40)
npm ERR!     at final (/usr/local/lib/node_modules/npm/node_modules/read-package-json/read-json.js:321:27)
npm ERR!     at /usr/local/lib/node_modules/npm/node_modules/read-package-json/read-json.js:139:33
npm ERR!     at cb (/usr/local/lib/node_modules/npm/node_modules/slide/lib/async-map.js:48:11)
npm ERR!     at /usr/local/lib/node_modules/npm/node_modules/read-package-json/read-json.js:316:40
npm ERR!     at fs.js:266:14
npm ERR!     at Object.oncomplete (fs.js:107:15)
npm ERR! If you need help, you may report this log at:
npm ERR!     <http://github.com/isaacs/npm/issues>
npm ERR! or email it to:
npm ERR!     <npm-@googlegroups.com>

npm ERR! System Darwin 14.3.0
npm ERR! command "node" "/usr/local/bin/npm" "install"
npm ERR! cwd /Volumes/SANDISK64/Workspace/optbot-org/ui
npm ERR! node -v v0.10.5
npm ERR! npm -v 1.2.18
npm ERR! 
npm ERR! Additional logging details can be found in:
npm ERR!     /Volumes/SANDISK64/Workspace/optbot-org/ui/npm-debug.log
npm ERR! not ok code 0
```

mscappini commented:
Works okay on my machine. However, I noticed you have some out of date node and npm packages? The latest version of Node is 0.12, and npm is 2.5. Looks like you are running 0.10 and 1.2.

aisthesis commented:
Thanks, the upgrade fixed it.

mscappini commented:
Glad to hear it!

mscappini commented:
@aisthesis I'm also okay with defining the "connects to" section as "Connects To" for now. I have no other immediate thoughts on that.

What's the status of this pull request now? What remains to be done before it can be merged?

aisthesis commented:
As soon as we're agreed on the documentation structure (my draft [here](https://github.com/optbot/template/pull/1)), I'd like to run it following your instructions. So:
1. Update README.md
2. I'll test on my VirtualBox VMI set up mirror the cloud environment.

Once it's merged and I have it going on the cloud itself, I'll also be interested to see if you guys can hit it using the cloud server's IP address. I may have to whitelist your IP addresses for that, though--which would make me feel better about our server's security!

aisthesis commented:
BTW I don't expect (as in `template`) that we can test that it's working by running `npm test`. But we should have a checklist of how to verify that the service is running properly. That is, there should be 'Testing' section, but the tests don't need to be automated.

Commit:
Moved code into lib/ directory.
Created a LICENSE file.
Updated package.json.

Commit:
Removed the startdev script. We'll save this for another time.

Commit:
Updated the readme.

aisthesis commented:
After installing with no errors, trying to run gives me this:

$ npm start

> @optbot/ui@0.0.1 start /mnt/disk1/home/marshall/workspace/ui
> node lib/app.js

module.js:340
    throw err;
          ^
Error: Cannot find module './config.production.js'
    at Function.Module._resolveFilename (module.js:338:15)
    at Function.Module._load (module.js:280:25)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at Object.<anonymous> (/mnt/disk1/home/marshall/workspace/ui/lib/config/config.js:2:18)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)

npm ERR! Linux 3.16.0-30-generic
npm ERR! argv "node" "/usr/local/bin/npm" "start"
npm ERR! node v0.10.25
npm ERR! npm  v2.7.6
npm ERR! code ELIFECYCLE
npm ERR! @optbot/ui@0.0.1 start: `node lib/app.js`
npm ERR! Exit status 8
npm ERR! 
npm ERR! Failed at the @optbot/ui@0.0.1 start script 'node lib/app.js'.
npm ERR! This is most likely a problem with the @optbot/ui package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node lib/app.js
npm ERR! You can get their info via:
npm ERR!     npm owner ls @optbot/ui
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /mnt/disk1/home/marshall/workspace/ui/npm-debug.log

mscappini commented:
Yeah, I just realized it would do that. It's missing a production config. I think I'll just make it ignore a config that doesn't exist and use all default values. Where default values are production values.

Unless you all think it wise we should have specific configs for specific environments? I've experienced pros and cons with each approach, but I'm mostly indifferent.

Or I could opt to use konphyg right now.

aisthesis commented:
Am I going to see a chart even if I don't have `restapi` installed? I do have mongo on my test environment with the dumped data I sent you.

mscappini commented:
No, the `ui` piece currently makes a REST call to the `restapi` which queries the database.

aisthesis commented:
Shouldn't we make sure that's working first then? Is that ready except for documentation?

Commit:
Updated the README with "Connects To".
Moved the Mongo connection string to config file.
Implemented nconf for configuration settings.

mscappini commented:
Please have another go on the server.

Commit:
Updated README to include a final example to change both host and port, and include default values for host and port.

Commit:
Fixed priority of loading configurations.

aisthesis commented on README.md:
```diff
@@ -4,22 +4,48 @@ Description
+
+        $ npm start [-- [--port=your_port] [--host=your_host]]
+
+2. If the service is running on port `8080` on a host with IP address `91.198.174.192`, navigate to `http://91.198.174.192:8080/` in a web browser.
```
This should be under 'Testing' along with a brief description of what to expect.

Commit:
Moved the testing piece of the instructions to a "Testing" section.

Commit:
Switched hardcoded expiry date to a time with multiple records.

Commit:
Standardizing text format.
Updated code to read configuration from npm config.
Updated README accordingly.

Commit:
Partial rendering supported.
The app now writes error messages to the console, opposed to a stack.
All response.render functions are being intercepted, and the 'req' and 'res' objects are being added to the context for use within template rendering.
A 'renderPartial' module was added to extend the response object and include a 'renderPartial' method that will look for a template in the partials directory.

aisthesis commented:
I hate to merge a non-functional version, but things have changed so much since this pull request that it will only be confusing to have all of this discussion when we get to the real pull request. So, I think it makes sense in this case to make an exception, particularly since we're not merging into a functioning master branch.

Go ahead and merge, but don't increment the version in the branch you're working on, since this is really still the same iteration.
